### ISSUE TITLE
Media field troubles when going headless

### ISSUE THREAD

**Issue by Markarend** [PiranhaCMS/piranha.core#1183]
Modifying Piranha for 'headless' scenario: I separated the api into its own REST API, and the MVC into its own UI that calls the REST API. Mostly it works well, but now an obstacle. My REST API serializes the result from the Piranha api, and my MVC web deserializes it into a Piranha StandardPage type. This works for all fields except **Media**, which is always null. Using newtonsoft.Json.

The **Media** property is defined in Piranha.Extend.Fields.MediaFieldBase with **internal set**, which explains why I can't deserialize into it. So I added a **[JsonProperty]** attribute to the **Media** property. Once I did this, then the **Media** field was correctly deserialized by the MVC, and images appeared.

But then I found this broke something else: in the manager, when I try to save a page with a Hero image, the Save buttons spins, stops as though it succeeded, but the toast never appears saying success. I set Debug logging for Microsoft.AspNetCore.Mvc.Infrastructure, and see there's a model state error when this happens:

```
dbug: Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker[2]
      Executed controller factory for controller Piranha.JA.Manager.Controllers.PageApiController (Piranha.JA.Manager)
dbug: Microsoft.AspNetCore.Mvc.Infrastructure.ModelStateInvalidFilter[1]
      The request has model state errors, returning an error response.
dbug: Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker[3]
      Request was short circuited at action filter 'Microsoft.AspNetCore.Mvc.Infrastructure.ModelStateInvalidFilter'.
```

I can debug into the PageApiController for most operations, but when Save executes none of the breakpoints are hit.

I also tried just removing the **internal** accessor from set (_and_ removing [JsonProperty]) and got the same behavior: can deserialize **Media** ok but trying to save page with a Hero from manager gives model state error.

Would love to write up some documentation on how to go headless if we can solve this.

We've been debugging this one for a few days, can anyone help?

**Comment by tidyui** [comment]
Hi there! As you can see in this issue changing the Media level to public or binding the media object like this in the manager has a history of crashing the manager.

https://github.com/PiranhaCMS/piranha.core/issues/1117

Now I can’t remember why this happens, because like you say it was very hard to pinpoint the error that occurred in the MVC model binding.

Which are the main properties you need from the media object?

**Comment by Markarend** [comment]
I need the media value so the image or other media can be rendered. It's simply null after deserializing, so no images are available to display. Thank you for linking to issue #1117, I'll read through that to familiarize myself with the issue.

**Comment by Markarend** [comment]
Reading through [#1117](https://github.com/PiranhaCMS/piranha.core/issues/1117) and [5a51cc8 ](https://github.com/PiranhaCMS/piranha.core/commit/5a51cc828646b75f6f0dde48b6d2a706ed6b014e) gave me an idea, and it looks like it's working for a few simple tests.  I added a property called MediaSerializable with a JsonProperty name of Media. This seems to solve both issues; I can save a Hero in the manager, and deserialize it in my MVC web:

```
        /// [JsonIgnore]
        public Models.Media Media { get; internal set; }

        [JsonProperty(PropertyName = "Media")]
        public Models.Media MediaSerializable {
            get {
                return Media;
            }
            set {
                Media = value;
            }
        }

```
The [JsonIgnore] doesn't seem to be necessary (yet).  Thanks for sparking this idea!

**Comment by tidyui** [comment]
The root issue is that when a **new** item is selected in the manager in the media picker, the media model that is assigned contains a formatted string size "xxx kb". When ASP.NET tries to deserialize this into a long it fails.

So changing the following line in the `.vue` components for the media-based fields:

    this.model.media = media;

to

    this.model.media = {
        id: media.id,
        folderId: media.folderId,
        type: media.type,
        filename: media.filename,
        contentType: media.contentType,
        publicUrl: media.publicUrl,
};

Resolves the serialization error in the manager when the `Media` property is public.. I'll push this fix to a new branch so you can test it.

Regards

**Comment by tidyui** [comment]
Please test with the branch https://github.com/PiranhaCMS/piranha.core/tree/features/make-mediafield-media-public

Regards

**Comment by tidyui** [comment]
Also, we're adding some stuff into our Api-module, like optional authorization on api-access. It would be great to know what API:s you've added and what you thing is needed for full coverage so we can update and add these into our standard api-package:

https://github.com/PiranhaCMS/piranha.core/tree/master/core/Piranha.WebApi

**Comment by Markarend** [comment]
Hakan, thank you SO MUCH for finding this root cause and giving a solution!  I tested the change you posted and it is working for me.  Note that I didn't pull your branch because we went "off the reservation" with our own copy, but made the change you showed in the 5 .vue files in manager. I also needed to retain my [JsonProperty] attribute on the Media field property, so my deserialization works, and with your fix, this attribute no longer breaks the manager. MUCH APPRECIATION!!
